Rework ChaCha_RNG

thillux opened the pull request:
When reading the source code of the ChaCha_RNG I saw some room for improvement:
- style: use initialization lists
- seeding: use HMAC(SHA-512) instead of HMAC(SHA-256) for less effects of the injectivity of hash functions
- internal state: add optional chacha key overwrite after each operation in order to provide backtracking resistance
- internal state: also use IV/nonce to widen effective key material

I'll adapted KAT test results via copy paste from CLI output and can also generate them externally if needed.

thillux commented:
CC @reneme

reneme commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -60,11 +73,18 @@ void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint
    m_hmac->update(input);
-   m_chacha->set_key(m_hmac->final());
+   const auto chacha_key = m_hmac->final();
+   m_chacha->set_key(chacha_key.data(), m_chacha_keylen);
```
As mentioned above: an assertion would be nice here. Maybe extract into a helper function and reuse it in both locations.

```C++
namespace {

std::span<const uint8_t> truncate(std::span<const uint8_t> in, size_t bytes) {
   BOTAN_ASSERT_NOMSG(in.size() >= bytes);
   return in.first(bytes);
}

}
```

```C++
void ChaCha_RNG::update(std::span<const uint8_t> input) {
   m_hmac->update(input);
   m_chacha->set_key(truncate(m_hmac->final(), m_chacha_keylen));
   // [...]
}
```

reneme commented on src/lib/rng/chacha_rng/chacha_rng.h:
```diff
@@ -116,6 +164,14 @@ class BOTAN_PUBLIC_API(2, 3) ChaCha_RNG final : public Stateful_RNG {
+      const static inline char* m_hmac_algo = "HMAC(SHA-512)";
+
+      // use "classic" 8 byte nonce as extended key material
+      const static inline size_t m_chacha_iv_len = 8;
```
Similarly as above: Please move to the top of the class and make it `constexpr size_t`

reneme reviewed (commented):
Just some general remarks so far.

thillux commented:
Ready from my side, will squash, when review is done.

thillux commented on src/lib/rng/chacha_rng/chacha_rng.h:
```diff
@@ -116,6 +163,11 @@ class BOTAN_PUBLIC_API(2, 3) ChaCha_RNG final : public Stateful_RNG {
+      const size_t m_chacha_keylen;
+
+      const static inline char* m_stream_cipher_algo = "ChaCha(20)";
+      const static inline char* m_hmac_algo = "HMAC(SHA-512)";
```
Done and moved.

reneme commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -60,16 +75,33 @@ void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint
+   // optionally overwrite key after each output operation for backtracking resistance
+   if(m_fast_key_erasure) {
+      // additional variable for key material is needed to stay on the stack
+      auto key_material = m_chacha->keystream_bytes(chacha_key_len + chacha_iv_len);
```
Nit:


```suggestion
      const auto key_material = m_chacha->keystream_bytes(chacha_key_len + chacha_iv_len);
```

reneme reviewed (commented):
@thillux I didn't realize I am authorized to push into your fork 😨 I meant to push into my own branch, but a sole `git push -f` was a bit too vigorous. Please make sure I didn't accidentally crush something. Sorry about that.

---

Changes itself look good to me and I think the improvements make sense. Thanks!

To the best of my understanding, the ChaCha RNG isn't standardized in any way, so we should have the liberty to improve the implementation as suggested. However, depending on how pedantically we want to follow semantic versioning, these changes might be seen as a violation. E.g. if someone based their unit tests on a hard-coded seed for this RNG their tests would break (much like Botan's KATs had to be updated).

As far as I can see, "breaking" changes are:

* "HMAC(SHA-256)" replaced by "HMAC(SHA-512)"
* derivation and usage of an IV for the internal ChaCha updates

@randombit What's your take on that? Do we want to guarantee stable DRNG outputs across minor versions?

thillux committed:
chacha_rng: add test for fast key erasure

This adapts the KATs for new key derivation with HMAC(SHA-512) and
optional fast key erasure.

Signed-off-by: Markus Theil <theil.markus@gmail.com>

thillux committed:
chacha_rng: small rework of key derivation and add fast key erasure

The current ChaCha_RNG implementation did not provide backtracking
resistance on internal state compromise and could insert more entropy
into the internal ChaCha20 state by additionally using the IV.

Change this by:
- Using HMAC(SHA-512) instead of HMAC(SHA-256) for key derivation
  on (re-)seed.
- Add optional rekeying after each output to add backtracking
  resistance.
- Also generate IV from input entropy.

This makes it possible, to have up to 320 bit of entropy in the
effective internal state (ChaCha key + IV).

A next commit will adapt KATs to this new structure.

Co-authored-by: René Meusel <github@renemeusel.de>
Signed-off-by: Markus Theil <theil.markus@gmail.com>

reneme commented on src/lib/rng/chacha_rng/chacha_rng.h:
```diff
@@ -20,24 +20,64 @@ class Entropy_Sources;
+*   implementation by many people and organizations.
+* - Providing optional fast key erasure is necessary to reach
+*   backtracking resistance of the internal state. This costs
+*   performance depending on the requests sizes of the user.
```
```suggestion
* - Providing optional fast key erasure is necessary to reach
*   backtracking resistance of the internal state. This costs
*   performance depending on the request sizes of the user.
```

reneme reviewed (approved):
Changes look good to me, apart from the semver concerns and a typo.

Oh, and please note and update [the related documentation](https://botan.randombit.net/handbook/api_ref/rng.html#chacha-rng) in `rng.rst`

thillux committed:
doc: update RNG documentation for ChaCha_RNG

Signed-off-by: Markus Theil <theil.markus@gmail.com>

thillux commented:
While there, I saw, that ESDM_RNG is not documented there. Can I do it in this PR or open another one?

reneme commented:
> Can I do it in this PR or open another one?

Please open a separate one.

Commit:
chacha_rng: add stability note in RNG doc

Signed-off-by: Markus Theil <markus.theil@secunet.com>

thillux commented:
@randombit I rebased and fixed compilation against recent master.

randombit commented on doc/api_ref/rng.rst:
```diff
@@ -190,21 +194,29 @@ for this use by NIST.
 future entropy additions), then the desired RNG outputs.
 
+Please note, that you should not rely on the stability of this RNG's outputs
+regarding identical seeds across versions. The output changed in version 3.10.0
```
Should be 3.13.0 here

thillux replied:
Fixed.

randombit commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -59,17 +69,28 @@ void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint
    }
 
+   //  Always generate next key material from the first blocks generated by a key
+   //  as it removes a positional argument from a possible analysis.
```
I'm having a difficult time parsing this comment

thillux replied:
I changed this comment. Please check if its clearer now. I can also drop this comment, if it adds nothing worth knowing.

randombit commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -8,48 +8,58 @@
+}
+
+void ChaCha_RNG::update_chacha_state(std::span<const uint8_t> key_material) {
+   BufferSlicer bs(key_material);
```
Just for sanity purposes this should assert that key_material.size() > chacha_key_len + chacha_iv_len, BufferSlicer will catch it at runtime but an assert is cleaner

thillux replied:
Added.

randombit commented on src/tests/test_rng_behavior.cpp:
```diff
@@ -673,20 +673,100 @@ class ChaCha_RNG_Unit_Tests final : public Stateful_RNG_Tests {
+         return result;
+      }
+
+      Test::Result test_max_number_of_bytes_per_request() override {
```
This and `test_reseed_interval_limits` below seem pretty pointless as tests

thillux replied:
They were there to satisfy the virtual base class. At least not duplicated now, by deriving from the ChaCha_RNG_Unit_Tests.

randombit commented on src/lib/rng/chacha_rng/chacha_rng.h:
```diff
@@ -60,8 +104,11 @@ class BOTAN_PUBLIC_API(2, 3) ChaCha_RNG final : public Stateful_RNG {
+      * on request size mix, deactivated by default
       */
-      BOTAN_FUTURE_EXPLICIT ChaCha_RNG(std::span<const uint8_t> seed);
+      BOTAN_FUTURE_EXPLICIT ChaCha_RNG(std::span<const uint8_t> seed, bool fast_key_erasure = false);
```
So recently we've started using named enums instead of bools and I think it leads to quite a bit more readable code vs raw `true`/`false`, eg a class-local `enum class KeyErasure : uint8_t { NoErasure, WithKeyErasure }`

thillux replied:
Done.

randombit reviewed (changes requested):
So change seems fine we can go ahead. A few minor comments. Also please rebase for CI purposes since at this point the base commit is pretty old and there may be some additional eg `clang-tidy` warnings that pop up

thillux commented:
On this now. Sorry for the delay.

thillux commented:
These CI failures seem to be unrelated to my changes:

```
[0](https://github.com/randombit/botan/actions/runs/31366759069/job/93386862253?pr=5121#step:6:5626)
  PASSED (ClientAuth-NoCertificate-Server-TLS-Sync)
  FAILED (TLS13-1RTT-Client-TLS-Sync)
  unexpected failure
  got:
  	stderr:
  		
  	local: "didResume is false, but we expected the opposite"
  	child: "none"
  	stdout: 
  
  extra stderr:
  
  PASSED (TLS13-HelloRetryRequest-Client-TLS-Sync)
  PASSED (TLS13-HelloRetryRequest-Server-TLS-Sync)
  PASSED (ClientAuth-ECDSA-Client-TLS-Sync)
  PASSED (ClientAuth-NoCertificate-Server-TLS13-TLS-Sync)
  ```
  
  Please let me know if I broke anything here.

thillux commented:
Updated comment for 3.14 and rebased to master.

randombit commented:
BoGo still failing, definitely needs some investigation as I have not seen BoGo to be flaky at any point. OTOH I definitely do *not* see how this change could affect BoGo or TLS at all....

randombit commented:
Oh interesting ... we are using `ChaCha_RNG` with a fixed seed for BoGo, possibly the change in behavior is introducing a flake, will investigate...

randombit commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -59,17 +71,32 @@ void ChaCha_RNG::generate_output(std::span<uint8_t> output, std::span<const uint
+   // the keystream (without updates in between). Therefore the next key becomes independent
+   // of the request pattern. This should not harm security and may ease mathematical analysis,
+   // as no variable position of the next key in the output stream has to be modeled.
+   const auto key_material = m_chacha->keystream_bytes(erase_key ? chacha_key_len + chacha_iv_len : 0);
```
This is going to require a heap allocation + deallocation plus a pretty unnecessary zeroization. I'd suggest storing instead to a local array (fixed at 32+8 = 40 bytes)

Also it calls into a virtual even if we're not using key erasure.

Something like this instead

```
std::array<uint8_t, chacha_key_len + chacha_iv_len> key_material{};
if(erase_key) {
  m_chacha->write_keystream(key_material);
}
```

randombit commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -8,48 +8,60 @@
+}
+
+void ChaCha_RNG::update_chacha_state(std::span<const uint8_t> key_material) {
+   BOTAN_ASSERT_NOMSG(key_material.size() >= chacha_key_len + chacha_iv_len);
```
Should probably be exact equality here, getting pass more implies something went wrong somewhere...

thillux replied:
This was not directly possible, as the sum gives 320 bit whereas SHA-512 returns 512 bit. I nevertheless tried to shave off the usage of the BufferSlicer. Please have a look again.

randombit commented on src/lib/rng/chacha_rng/info.txt:
This date should be updated to indicate the changed design

thillux replied:
Done.

randombit commented on src/lib/rng/chacha_rng/chacha_rng.cpp:
```diff
@@ -8,48 +8,60 @@
+
+   BufferSlicer bs(key_material);
+   m_chacha->set_key(bs.take<chacha_key_len>());
+   m_chacha->set_iv(bs.take<chacha_iv_len>());
```
If the span is always checked to be exactly the correct length this could skip the BufferSlicer entirely and use `key_material.first(chacha_key_len)` and `key_material.last(chacha_iv_len)`

thillux replied:
Done.

randombit reviewed (changes requested):
Very close, a few more nits. Also needs a rebase to pick up the BoGo fix

thillux commented:
@randombit Should be done now. Please have a look again at the assert + BufferSlice removal. As this was not directly possible as you suggested.

thillux commented:
Rebased, to fix the conflict in news.rst.
